slkspec/core.py

```diff
@@ -1,129 +1,355 @@
+                          slk_cache="/scratch/b/b12346").open()
+        dset = xr.open_dataset(url)
+
+    """
+
+    write_msg: str = "Write mode is not suppored"
+    """Error message that is thrown if the files are attempted to be opened
+    in any kind of write mode."""
+
+    def __init__(
+        self,
+        url: str,
+        local_file: str,
+        *,
+        override: bool = True,
+        mode: str = "rb",
+        touch: bool = True,
+        file_permissions: int = 0o3777,
+        _lock: threading.Lock = _retrieval_lock,
+        _file_queue: Queue[Tuple[str, str]] = FileQueue,
+        **kwargs: Any,
+    ):
+        if not set(mode) & set("r"):  # The mode must have a r
+            raise NotImplementedError(self.write_msg)
+        if "b" not in mode:
+            kwargs.setdefault("encoding", "utf-8")
+        self._file = str(Path(local_file).expanduser().absolute())
+        self._url = str(url)
+        self.touch = touch
+        self.file_permissions = file_permissions
+        self._order_num = 0
+        self._file_obj: Optional[IO[Any]] = None
+        self._lock = _lock
+        self.kwargs = kwargs
+        self.mode = mode
+        self.newlines = None
+        self.error = "strict"
+        self.encoding = kwargs.get("encoding")
+        self.write_through = False
+        self._file_queue = _file_queue
+        print(self._file)
+        with _lock:
+            if not Path(self._file).exists() or override:
+                self._file_queue.put((self._url, str(Path(self._file).parent)))
+            elif Path(self._file).exists():
+                if self.touch:
+                    Path(self._file).touch()
+                self._file_obj = open(self._file, mode, **kwargs)
+
+    @property
+    def name(self) -> str:
+        """Get the file for the SLKFile object."""
+        if self._file_obj is not None:
+            return self._file
+        return self._url
+
+    def _retrieve_items(self, retrieve_files: list[tuple[str, str]]) -> None:
+        """Get items from the tape archive."""
+
+        retrieval_requests: Dict[Path, List[str]] = defaultdict(list)
+        logger.critical("Retrieving %i items from tape", len(retrieve_files))
+        for inp_file, out_dir in retrieve_files:
+            retrieval_requests[Path(out_dir)].append(inp_file)
+        for output_dir, inp_files in retrieval_requests.items():
+            output_dir.mkdir(parents=True, exist_ok=True, mode=self.file_permissions)
+            logger.critical("Creating slk query for %i files", len(inp_files))
+            search_str = pyslk.slk_search(pyslk.slk_gen_file_query(inp_files))
+            search_id_re = re.search("Search ID: [0-9]*", search_str)
+            if search_id_re is None:
+                raise ValueError("No files found in archive.")
+            search_id = int(search_id_re.group(0)[11:])
+            logger.critical("Retrieving files for search id: %i", search_id)
```

Comment:
Prior to this line, I would add a check on how many results were found. 

(a) `slk search` returns as search_id even if no files were found. We have a feature request to StrongLink: "slk search: not return a search_id if no results are found". But, this feature request has a low priority and I don't know when/if it will come.

(b) Currently, `pyslk.slk_list` throws an error if it receives a search_id with no search results. This is so far OK (from my point of view) because the `pyslk.slk_*` wrappers should be dumb wrappers. However, there is a `parsers.slk_list_formatted` which returns a Pandas.DataFrame. This should handle the error properly and return an empty DataFrame. I'll try to fix this today evening. Question to @antarcticrainforest  would be: Do you want to catch the error of `pyslk.slk_list` or do you want to check the output of `parsers.slk_list_formatted` for emptiness?

The error message in the first case would be: `ERROR: No resources found for given search id: 246936` (if search_id was 246936). It would be this error in total `pyslk.exceptions.PySlkException: pyslk.slk_list: ERROR: No resources found for given search id: 246936`